sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala

```diff
@@ -794,6 +802,114 @@ class SparkSession private(
     }
   }
 
+
+  /**
+   * Add a tag to be assigned to all the operations started by this thread in this session.
+   *
+   * Often, a unit of execution in an application consists of multiple Spark executions.
+   * Application programmers can use this method to group all those jobs together and give a group
+   * tag. The application can use `org.apache.spark.sql.SparkSession.interruptTag` to cancel all
+   * running executions with this tag. For example:
+   * {{{
+   * // In the main thread:
+   * spark.addTag("myjobs")
+   * spark.range(10).map(i => { Thread.sleep(10); i }).collect()
+   *
+   * // In a separate thread:
+   * spark.interruptTag("myjobs")
+   * }}}
+   *
+   * There may be multiple tags present at the same time, so different parts of application may
+   * use different tags to perform cancellation at different levels of granularity.
+   *
+   * @param tag
+   *   The tag to be added. Cannot contain ',' (comma) character or be an empty string.
+   *
+   * @since 4.0.0
+   */
+  def addTag(tag: String): Unit = {
+    SparkContext.throwIfInvalidTag(tag)
+    userDefinedToRealTagsMap.put(tag, s"spark-session-$sessionUUID-$tag")
+  }
+
+  /**
+   * Remove a tag previously added to be assigned to all the operations started by this thread in
+   * this session. Noop if such a tag was not added earlier.
+   *
+   * @param tag
+   *   The tag to be removed. Cannot contain ',' (comma) character or be an empty string.
+   *
+   * @since 4.0.0
+   */
+  def removeTag(tag: String): Unit = {
+    SparkContext.throwIfInvalidTag(tag)
+    userDefinedToRealTagsMap.remove(tag)
+  }
+
+  /**
+   * Get the tags that are currently set to be assigned to all the operations started by this
+   * thread.
+   *
+   * @since 4.0.0
+   */
+  def getTags(): Set[String] = userDefinedToRealTagsMap.keys().asScala.toSet
+
+  /**
+   * Clear the current thread's operation tags.
+   *
+   * @since 4.0.0
+   */
+  def clearTags(): Unit = userDefinedToRealTagsMap.clear()
+
+  /**
+   * Interrupt all operations of this session that are currently running.
+   *
+   * @return
+   *   sequence of Job IDs of interrupted operations.
+   * @since 4.0.0
+   */
+  def interruptAll(): Seq[String] = {
+    val cancelledIds = sparkContext.cancelAllJobs(_.getSparkSessionUUID.contains(sessionUUID))
+    ThreadUtils.awaitResult(cancelledIds, 60.seconds).map(_.toString).toSeq
+  }
+
+  /**
+   * Interrupt all operations of this session with the given operation tag.
+   *
+   * @return
+   *   sequence of Job IDs of interrupted operations.
+   *
+   * @since 4.0.0
+   */
+  def interruptTag(tag: String): Seq[String] = {
```

Comment:
Interrupt API is non-blocking IIRC, also it should work with streaming queries too.

Reply:
Oh good to know, I think it's possible to invoke the callback (with Job IDs which are going to be cancelled) earlier to make it non-blocking.

Reply:
Done! Now the scheduler will first collect job IDs that will be cancelled, then unblock the caller, and then process the cancellation.